File: Practice/Recursion/AlgoExpert/NthFibonaci/solution2.cpp
using namespace std;

unordered_map<int, int> memoize;
memoize[1]=0;
memoize[2]=1;

int getNthFib(int n, unordered_map<int, int> memoize) {
	
  // Write your code here.
	if(memoize.find(n)){
		return memoize[n];
	}else{
		memoize[n] = getNthFib(n-1, memoize) + getNthFib(n-2, memoize);
		return memoize[n];
	}
}

File: Basic/AlgoExpert/StairCaseTransversal/solution2.py
def staircaseTraversal(height, maxSteps):
    # Write your code here.
	return numberOfWaysToTheTop(height, maxSteps, {0:1, 1:1})
    
def numberOfWaysToTheTop(height, maxSteps, memoize):
	if height in memoize:
        return memoize[height]
	
	numberOfWays = 0
	for step in range(1, min(height, maxSteps)+1):
		numberOfWays += numberOfWaysToTheTop(height-step, maxSteps)
    
    memoize[height] = numberOfWays
		
	return numberOfWays

# Time Complexity O(K^N)  | Space Complexity O(n)